File: src/runtime.cpp
#include "gpu_runtime/fusion.hpp"
#include "gpu_runtime/inference.hpp"
#include "gpu_runtime/io.hpp"
#include "gpu_runtime/replay.hpp"
#include "gpu_runtime/sync.hpp"
#include "gpu_runtime/threadsafe_queue.hpp"
#include "gpu_runtime/types.hpp"

#include <deque>
#include <iostream>
#include <optional>
#include <vector>

namespace gpu_runtime {

Summary run_pipeline(const RunOptions& options) {
  const auto camera_packets = load_camera_packets(options.data_dir);
  const auto imu_packets = load_imu_packets(options.data_dir);

  ThreadSafeQueue<ReplayMessage<CameraPacket>> camera_queue;
  ThreadSafeQueue<ReplayMessage<IMUPacket>> imu_queue;

  ReplayThread<CameraPacket> camera_thread(camera_packets, options.replay_speed, camera_queue);
  ReplayThread<IMUPacket> imu_thread(imu_packets, options.replay_speed, imu_queue);

  camera_thread.start();
  imu_thread.start();

  SyncCoordinator coordinator(options.sync_tolerance_ms);
  auto backend = create_inference_backend(options.device);
  EventFuser fuser;
  std::vector<FusedEvent> events;
  std::deque<CameraPacket> pending_cameras;

  bool camera_done = false;
  bool imu_done = false;

  auto process_pending = [&]() {
    std::deque<CameraPacket> still_pending;
    while (!pending_cameras.empty()) {
      CameraPacket cam = pending_cameras.front();
      pending_cameras.pop_front();
      auto synced = coordinator.try_sync_camera(cam);
      if (!synced.has_value()) {
        still_pending.push_back(cam);
        continue;
      }
      auto inference = backend->infer(*synced);
      auto event = fuser.fuse(*synced, inference);
      events.push_back(event);
    }
    pending_cameras = std::move(still_pending);
  };

  while (!(camera_done && imu_done)) {
    if (!imu_done) {
      auto msg = imu_queue.pop();
      if (msg.end_of_stream) {
        imu_done = true;
      } else {
        coordinator.ingest_imu(msg.payload);
        process_pending();
      }
    }

    if (!camera_done) {
      auto msg = camera_queue.pop();
      if (msg.end_of_stream) {
        camera_done = true;
      } else {
        pending_cameras.push_back(msg.payload);
        process_pending();
      }
    }
  }

  process_pending();

  camera_thread.join();
  imu_thread.join();

  Summary summary = write_outputs(events, options.output_dir);
  std::cout << "--- Replay Summary ---\n";
  std::cout << "device: " << summary.device << "\n";
  std::cout << "frames_processed: " << summary.frames_processed << "\n";
  std::cout << "aligned_pairs: " << summary.aligned_pairs << "\n";
  std::cout << "average_sync_gap_ms: " << summary.average_sync_gap_ms << "\n";
  std::cout << "average_inference_latency_ms: " << summary.average_inference_latency_ms << "\n";
  std::cout << "p95_inference_latency_ms: " << summary.p95_inference_latency_ms << "\n";
  std::cout << "high_risk_events: " << summary.high_risk_events << "\n";
  std::cout << "medium_risk_events: " << summary.medium_risk_events << "\n";
  std::cout << "low_risk_events: " << summary.low_risk_events << "\n";

  return summary;
}

}  // namespace gpu_runtime

File: src/gpu_sensor_runtime/runtime.py
from __future__ import annotations

import queue
from collections import deque
from pathlib import Path
from typing import Any

from src.gpu_sensor_runtime.fusion.fuser import EventFuser
from src.gpu_sensor_runtime.inference.backends import TorchHeuristicInferenceBackend
from src.gpu_sensor_runtime.observability.metrics import MetricsCollector, write_summary, write_timeline
from src.gpu_sensor_runtime.replay.loaders import ReplayThread, StreamSentinel, load_camera_packets, load_imu_packets
from src.gpu_sensor_runtime.replay.synchronizer import SyncCoordinator
from src.gpu_sensor_runtime.types import CameraPacket, FusedEvent, IMUPacket


def run_pipeline(
    data_dir: Path,
    output_dir: Path,
    device: str = "auto",
    sync_tolerance_ms: float = 30.0,
    replay_speed: float = 20.0,
) -> dict[str, Any]:
    camera_packets = load_camera_packets(data_dir)
    imu_packets = load_imu_packets(data_dir)

    camera_queue: queue.Queue = queue.Queue()
    imu_queue: queue.Queue = queue.Queue()

    camera_thread = ReplayThread(camera_packets, camera_queue, "camera", replay_speed)
    imu_thread = ReplayThread(imu_packets, imu_queue, "imu", replay_speed)

    camera_thread.start()
    imu_thread.start()

    coordinator = SyncCoordinator(tolerance_ms=int(sync_tolerance_ms))
    backend = TorchHeuristicInferenceBackend.create(device=device)
    fuser = EventFuser()
    metrics = MetricsCollector()
    events: list[FusedEvent] = []

    imu_done = False
    camera_done = False
    pending_cameras = deque()

    def _process_pending() -> None:
        nonlocal pending_cameras
        keep = deque()
        while pending_cameras:
            cam = pending_cameras.popleft()
            synced = coordinator.try_sync_camera(cam)
            if synced is None:
                keep.append(cam)
                continue
            inference = backend.infer(synced)
            event = fuser.fuse(synced, inference)
            metrics.record(event, backend.device)
            events.append(event)
        pending_cameras = keep

    while not (imu_done and camera_done):
        try:
            imu_item = imu_queue.get(timeout=0.01)
            if isinstance(imu_item, StreamSentinel):
                imu_done = True
            elif isinstance(imu_item, IMUPacket):
                coordinator.ingest_imu(imu_item)
                _process_pending()
        except queue.Empty:
            pass

        try:
            cam_item = camera_queue.get(timeout=0.01)
            if isinstance(cam_item, StreamSentinel):
                camera_done = True
            elif isinstance(cam_item, CameraPacket):
                pending_cameras.append(cam_item)
                _process_pending()
        except queue.Empty:
            pass

    _process_pending()

    camera_thread.join(timeout=1.0)
    imu_thread.join(timeout=1.0)

    summary = metrics.summary()
    write_timeline(events, output_dir)
    write_summary(summary, output_dir)
    return summary
